### PR TITLE
Add DDPM model unit tests

### LINKED ISSUE #83: Add tests for DDPM model
### Problem / Goal

The DDPM implementation currently lacks unit tests. To align with the testing style in the main pipeline, we should add tests to cover different cases.

### What does good look like?

- [ ] A new test for DDPM is added.
- [ ] Tests follow the existing style in the main pipeline.

### Related context

_No response_

### PR DESCRIPTION
**Pull request by sylvesterkaczmarek**
Closes alan-turing-institute/icenet-mp#83

### CODE REVIEW

File: tests/models/test_ddpm.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    @pytest.mark.parametrize("use_autoregressive", [True, False])
+    def test_training_step_shapes(
+        self,
+        cfg_loss: DictConfig,
+        monkeypatch: pytest.MonkeyPatch,
+        use_autoregressive: bool,
+    ) -> None:
+        model = _make_model(cfg_loss, use_autoregressive=use_autoregressive)
+        model.model = _ZeroDenoiser()
```
Comment (jemrobinson) [inline tests/models/test_ddpm.py]: For this to work, you need to type-hint `DDPM.model` as `nn.Module`

File: tests/models/test_ddpm.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        conditioning = model.prepare_inputs(_make_batch())
+
+        assert conditioning.shape == (2, model.cond_channels, 16, 16)
+
+    @pytest.mark.parametrize("use_autoregressive", [True, False])
+    def test_training_step_shapes(
+        self,
+        cfg_loss: DictConfig,
+        monkeypatch: pytest.MonkeyPatch,
+        use_autoregressive: bool,
```
Comment (jemrobinson) [inline tests/models/test_ddpm.py:63]:
```suggestion
        *,
        use_autoregressive: bool,
```

File: tests/models/test_ddpm.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    ) -> None:
+        model = _make_model(cfg_loss, use_autoregressive=use_autoregressive)
+        model.model = _ZeroDenoiser()
+        monkeypatch.setattr(model, "log", lambda *_args, **_kwargs: None)
```
Comment (jemrobinson) [inline tests/models/test_ddpm.py]: Is this needed? I think log messages should be consumed by pytest unless running with caplog.

File: tests/models/test_ddpm.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    def forward(
+        self,
+        noisy: torch.Tensor,
+        _timesteps: torch.Tensor,
+        _conditioning: torch.Tensor,
+    ) -> torch.Tensor:
+        return torch.zeros_like(noisy)
+
+
+def _make_model(cfg_loss: DictConfig, *, use_autoregressive: bool = True) -> DDPM:
```
Comment (jemrobinson) [inline tests/models/test_ddpm.py]: The style in this repo is to make these helper functions member functions of `TestDDPM`

File: tests/models/test_ddpm.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        output_space={"channels": 1, "name": "osisaf-north", "shape": (16, 16)},
+        optimizer={},
+        scheduler={},
+        start_out_channels=4,
+        timesteps=4,
+        use_autoregressive=use_autoregressive,
+    )
+
+
+def _make_batch(batch_size: int = 2) -> dict[str, torch.Tensor]:
```
Comment (jemrobinson) [inline tests/models/test_ddpm.py]: Same as above.

**Review by jemrobinson — CHANGES_REQUESTED** [review]
Thanks for this PR @sylvesterkaczmarek ! I've got a couple of a comments on what you've added here.

I don't think this closes #83 as-is since the following things aren't covered here

- `validation_step()` / `test_step()`
- `sample()` / `_sample_parallel()` / `_sample_autoregressive()` reverse diffusion loop
- autoregressive sliding-window logic in `_sample_autoregressive()`
- `Mask/RestrictRange` (although these could be tested elsewhere)
- the `GaussianDiffusion`/`UNetDiffusion` modules
- `q_sample`, `p_sample`, `calculate_v` are only tested through the fake denoiser

**Review by jemrobinson — APPROVED** [review]
LGTM. Thanks @sylvesterkaczmarek !
